docs: ADR for graduating EventBridge in CAPA (#3583)
kubernetes-sigs/cluster-api-provider-aws: docs/adr/0005-graduation of EventBridge.md

# 5. Graduation of EventBridge in CAPA

* Status: [proposed]
* Date: TBD 
* Authors: @Ankitasw
* Deciders: @richardcase @sedefsavas

## Context
EventBridge is an eventbus provided by AWS to watch for the events generated by any SaaS application running on AWS, or any of the AWS services, that could be used by any other services.

> With respect to CAPA, EventBridge helps watch EC2 instance state change events using which _AWSMachine_ controller would trigger reconciliation and handle the lifecycle of the EC2 instances.

Reviewer: ```suggestion
With respect to CAPA, the  _AWSMachine_ controller uses EC2 instance state change events via EventBridge to  trigger reconciliation so that it can handle the lifecycle of the EC2 instances.
```

> There are other `CloudWatch` events that exists which can be watched upon and CAPA controllers can take necessary actions based on the events, for eg. ASG Scale-in events, Spot Instance Termination Notices, Scheduled Maintenance events etc.

Reviewer: ```suggestion
In the future, there are other `CloudWatch` events that CAPA controllers might want to take action on based on the event (e.g.  ASG Scale-in events, Spot Instance Termination Notices, Scheduled Maintenance events).
```

> ## Decision

Reviewer: An ADR is used to record a lightweight decision and the consequences of this decision. I'm struggling to understand what is the decision here from the description.

For me the decision is "We will graduate the EventBridge support out of experimental so that it is GA available and enabled by default in CAPA because we want to be able to use the different event types in the future."

Author: Makes sense to me, Will add this in decision

This document describes the ways on how to accommodate watching different types of events and how controllers could utilize these events. The scenarios that would need modification in EventBridge in CAPA (in `AWSInstanceState` controller) are as below:
* In the current _AWSMachinePool_ implementation, we don't have node draining logic. With the use of lifecycle hooks, CAPA controllers need to watch `ASG scale-in` events and act on them such that the EC2 instances shuts down gracefully.
* We also plan to support spot instances with _AWSMachinePool_ which would require CAPA controllers to watch for `Spot Instance Termination Notices(SITN)` which would give enough time to drain spot instances gracefully before they shut down.
* Handle scheduled maintenance events where the EC2 instance needs to be restarted because of reasons like hardware/software updates.

The `AWSInstanceState` controller would accommodate different types of events and would be flexible enough to be used across CAPA controllers.

## Consequences

> ### Current Design

Reviewer: I don't we should be covering "current design" in the consequences section. We really only need to cover the what will change as a result of this decision....so the stuff in the "proposed design" section.

Author: Shall we add this in decision section, or create a separate section?

Reviewer: I think there is too much detail overall. I think the decision section only needs to be a few sentances max.

Author: Since this is a bigger change, I wanted to highlight design decision as well, so if I want to add the design part, can I add a separate section in ADR only? Otherwise if I directly raise PR, it would be too huge to understand. But since this was not that bigger feature change, I thought of adding this to ADR, wdyt @richardcase ?

Author: or else i create a separate google doc and just reference it in decision section.

Reviewer: In that case, yes let's add the detail in the "current design" to the context.

Author: and proposed design in decision section only?

Reviewer: Proposed design in the consequences. Keep the decision small.

Author: sure, thanks for the inputs @richardcase 🙂

EventBridge looks for EC2 state change events for all the _AWSMachine_ based on `InstanceID` found in the SQS queue, and loads the state to `ec2-instance-state` label in that _AWSMachine_ in `processMessage()` func.

Currently, EventBridge handles messages only from source `aws.ec2`.

### Proposed Design
* We would define `EventBridgeEvent` struct in `awsinstancestate_controller.go` to capture EventBridge event details:

```go

> type EventBridgeEvent struct {

Reviewer 2: Is there a `type` field associated with the events to show the source? 
I am assuming `detail-type` is not showing the source as aws.ec2 or aws.autoscaling.

Reviewer 3: The `source` field will contain the context of what service sent the event, i.e., `aws.ec2` or `.aws.autoscaling`. 
The contents of `detail-type` can vary by service and type of event. A reference of events from AWS services can be found [here](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-service-event.html). For example, here's an EC2 Instance-terminate lifecycle event from EventBridge (from [EC2 Auto Scaling ref](https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-event-reference.html)):

```
{
  "version": "0",
  "id": "12345678-1234-1234-1234-123456789012",
  "detail-type": "EC2 Instance-terminate Lifecycle Action",
  "source": "aws.autoscaling",
  "account": "123456789012",
  "time": "yyyy-mm-ddThh:mm:ssZ",
  "region": "us-west-2",
  "resources": [
    "auto-scaling-group-arn"
  ],
  "detail": { 
    "LifecycleActionToken":"87654321-4321-4321-4321-210987654321", 
    "AutoScalingGroupName":"my-asg", 
    "LifecycleHookName":"my-lifecycle-hook", 
    "EC2InstanceId":"i-1234567890abcdef0", 
    "LifecycleTransition":"autoscaling:EC2_INSTANCE_TERMINATING", 
    "NotificationMetadata":"additional-info"
  } 
}
```

Author: @sedefsavas This detail is captured in line 49.
